### PR TITLE
Fix S2178 Rule message: Mention extracting right operand if applicable

### LINKED ISSUE #8559: Fix S2178 Rule message: Mention extracting right operand if applicable
S2178's message as defined in RSPEC:
`Correct this "x" to "xx" [and extract the right operand to a variable if it should always be evaluated].`
Our analyzer does not implement the optional part. This can lead developers into a trap, changing the behavior of their code without noticing.
Our analyzer should detect if the right operand is constant or a pure method, and if it is not, use the extended message.

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #8559

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs:57]: An alternative would be to look for side effects in the right hand side:
* invocation (`Count() > 5`)
* increment/decrement (`i++ > 5`)
* ...

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UseShortCircuitingOperator.cs
Comment (antonioaversa) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/UseShortCircuitingOperator.cs:36]: If we don't care about the result of this private method, why not always return `true` or `false`? Having `DateTime.Now` in a test makes the test non-deterministic. While the result of the test may not become itself flaky, I would still avoid any source of non-determinism.

Also, `DateTime.Now` can raise an exception in some "hostile contexts": e.g. `TimeZoneNotFoundException`.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs
Comment (antonioaversa) [inline analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs:55]: Ok for `ILocalSymbol`, `IFieldSymbol` and `IParameterSymbol`.
As for `IPropertySymbol`: I guess the assumption is that the getter "plays fair", and doesn't produce side effects, such as raising exceptions or validating preconditions.
Is this the rationale behind adding `IPropertySymbol` here?

File: analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs
Comment (antonioaversa) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs:57]: Yes. I think there is no need to go further for the time being.
We will maybe reconsider in the future, based on community inputs on the rule.

**Review by antonioaversa — CHANGES_REQUESTED** [review]
Implementation is good, in my opinion.
I have a few (minor, mostly stylistic) remarks and some questions, before this gets merged.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs:48]: I use pattern matching for assigning a new value all the time (even if it always returns true). This makes code much more readable especially if the variable gets introduced in the middle of an `&&` sequence, like with the `BinaryExpressionLeft(node) is { } left` which is also always true.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UseShortCircuitingOperator.cs
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/UseShortCircuitingOperator.cs:36]: I want to make sure, that interprocedural code detection is not able to determine the result at compile time. I used `Environment.Is64BitOperatingSystem` instead now.

File: analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/UseShortCircuitingOperatorBase.cs:55]: IPropertySymbol getters are already presumed to play nice, e.g. with pattern matching. As such, we can assume the same here.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/UseShortCircuitingOperator.cs
Comment (antonioaversa) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/UseShortCircuitingOperator.cs]: I have also just realized that we are not running that test code really :-)
All good

**Review by antonioaversa — APPROVED** [review]
Approving with 1 minor conventional remark added.
Please, have a look at the coverage, that is currently reported at 87.5%.

**Comment by antonioaversa** [comment]
@martin-strecker-sonarsource Coverage is back. There is a single line uncovered: https://sonarcloud.io/component_measures?metric=new_coverage&selected=sonaranalyzer-dotnet%3Aanalyzers%2Fsrc%2FSonarAnalyzer.Common%2FRules%2FUseShortCircuitingOperatorCodeFixBase.cs&view=list&pullRequest=8598&id=sonaranalyzer-dotnet

**Comment by martin-strecker-sonarsource** [comment]
Peach validation: No changes (as expected).

**Pull request merged**
